view_update_generator: Increase the registration_queue_size

asias opened the pull request:
When repair writes a sstable to disk, we check if the sstable needs view update processing. If yes, the sstable will be registered into the view update processing logic with the _registration_sem semaphore. However, the semaphore is not very useful because it is too late to add the back pressure since the sstable is already written.

We have seen multiple cases in the field where the _registration_sem is stuck so that repair is stuck since repair could not register any more sstables for view update processing. Or view update processing is too slow which blocks repair to finish on time.

This patch increases the registration_queue_size to a big number so that the view update processing won't block repair in practice. A follow up patch is planned to remove the semaphore completely. This patch is a simpler version for easier backport to fix issue in the field.

denesb commented on db/view/view_update_generator.hh:
```diff
@@ -50,7 +50,7 @@ using wait_for_all_updates = bool_class<wait_for_all_updates_tag>;
 class view_update_generator : public async_sharded_service<view_update_generator> {
 public:
-    static constexpr size_t registration_queue_size = 5;
+    static constexpr size_t registration_queue_size = 1000;
```
If you want to disable this for good, init it to maxing, not just some random high number.

denesb reviewed (commented):
I am not sure this is the direction we should go, in the past we talked about going in the opposite direction, e.g. this PR conflicts directly with https://github.com/scylladb/scylladb/issues/11743.

Generating view updates for the sstables written by repair is part of the repair itself. This PR will effectively result in repair being declared done, when part of it is still running in the background.
Also, sstables written by repair will not be compacted while view generation is ongoing, this can easily OOM the node in the extreme case.

nyh commented:
> We have seen multiple cases in the field where the _registration_sem is stuck so that repair is stuck since repair could not register any more sstables for view update processing. Or view update processing is too slow which blocks repair to finish on time.

I'm not sure I understand the problem you are solving. View updates is indeed a very slow process when a repair has a lot of **write** work (if the repair just compares data and finds no inconsistencies, MV don't slow it down). The problem is whereas normal writes that repair does are trivial (just write sequentially to a sstable!), they are an order of magnitude slower when a view exists - to do the write you first need to do a random-access-read, and then also send the view updates. So we more-or-less **expect** that a repair that finds a lot of rows to fix will be much slower with MV, and we can't fix that.

I seems you want the repair itself to return quickly but do the MV updates lazily - potentially for many many hours after the repair "completed". But why does this make sense? would real users want that? How is this different from a repair of some table pretending to finish after it really finished repairing only half of it, and then continue to fix the second half in the background?

Moreover, as other reviewers noted, wouldn't having all these "staging" sstables not being compacted cause problems reading from the base table?

You mentioned that the repair was "stuck". Was it really "stuck" or just slow as expected?

asias commented:
> > We have seen multiple cases in the field where the _registration_sem is stuck so that repair is stuck since repair could not register any more sstables for view update processing. Or view update processing is too slow which blocks repair to finish on time.
> 
> I'm not sure I understand the problem you are solving. View updates is indeed a very slow process when a repair has a lot of **write** work (if the repair just compares data and finds no inconsistencies, MV don't slow it down). The problem is whereas normal writes that repair does are trivial (just write sequentially to a sstable!), they are an order of magnitude slower when a view exists - to do the write you first need to do a random-access-read, and then also send the view updates. So we more-or-less **expect** that a repair that finds a lot of rows to fix will be much slower with MV, and we can't fix that.
> 
> I seems you want the repair itself to return quickly but do the MV updates lazily - potentially for many many hours after the repair "completed". But why does this make sense? would real users want that? How is this different from a repair of some table pretending to finish after it really finished repairing only half of it, and then continue to fix the second half in the background?

Yes. I want to decouple repair and view update , so repair and view update processing can work independently. And make view update processing  a true background process which picks sstables in the staging directory and processes them. 

Note, currently, repair finishes only when the sstable is added to the view update queue. It does not wait for view update to finish to announce the repair has finished.  

The difference is that base table repair has really finished, not pretending. Say without waiting for view update to finish, repair finishes in 1 day,  with waiting for view update to finish, repair finishes in 11 days.  We could start dropping massive tombstones 10 days earlier.  We could have better read perf for base table. Or if the gc grace seconds is 10 days, we could avoid data resurrection for base table. 
 
> 
> Moreover, as other reviewers noted, wouldn't having all these "staging" sstables not being compacted cause problems reading from the base table?

I think this is not a big problem. The sstables generated by repair are disjoint and each sstable covers a vnode token range. 

> 
> You mentioned that the repair was "stuck". Was it really "stuck" or just slow as expected?

For example, we saw for instance the view_update_generator_pending_registrations metrics stayed a constant number for days.  But of course, this patch does not solve the stuck of the _registration_sem issue.  I am mentioning this because I want to remove the  _registration_sem completely eventually.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1813/

denesb commented:
> I think this is not a big problem. The sstables generated by repair are disjoint and each sstable covers a vnode token range.

It is a big problem. Each sstable has an in-memory footprint and above a certain number, just the sheer number of sstables can OOM a node.

denesb commented:
> So the problem is that the number of sstables is unbounded?
> 
> Perhaps we should defer starting the repair if we have backlog from the previous repair.

Yes. Or maybe we can use our new task framework to track the progress of view update generation and add the ability for the admin to wait for it.

avikivity commented:
Let's reject repairs if there is backlog. Relying on the administrator is not deterministic.

avikivity commented:
There's no point to compact disjoint sstables. It's just busy-work.

vladzcloudius commented:
> There's no point to compact disjoint sstables. It's just busy-work.

There is a very good reason to compact tiny disjoint sstables, @avikivity. The make them one large sstable and avoid scylla from OOMing while not making the repair ridiculously slow.

If it's important to keep sstables disjoint after this compaction - an LCS compaction can be used for that.

avikivity commented:
A large number of sstables doesn't cause an OOM, unless it's ridiculously large. A few thousand sstables should be fine.

raphaelsc commented:
@vladzcloudius I don't understand why you think compaction of staging will make things any better. We're not deduplicating data, just reducing the sstable count. Could you please expand a bit more why you think that staging should be compacted? If that's to avoid the throttle slowing down repair, then it's a problem with throttling heuristic. We won't reshape the newly added sstables just to satisfy the existing weak heuristic.

raphaelsc commented:
@asias https://github.com/scylladb/scylladb/pull/14179/commits/fb95ff4efac8efa272a90a9a0ae1eb6c01a42963 this sounds like an important fix, but the commit message made it hard for me to understand the consequences exactly. could you please expand a bit more on that and tell me if it merits a backport?

vladzcloudius commented:
> @vladzcloudius I don't understand why you think compaction of staging will make things any better. We're not deduplicating data, just reducing the sstable count. Could you please expand a bit more why you think that staging should be compacted? If that's to avoid the throttle slowing down repair, then it's a problem with throttling heuristic. We won't reshape the newly added sstables just to satisfy the existing weak heuristic.

According to @bhalevy the main reason of throttling was to avoid aggregation of too many small staging sstables. And we know that scylla doesn't handle such case well at all - there were multiple issues about that (and we were talking hundreds of thousands of sstables, @avikivity). 

In addition having too many small sstables is indeed going to create a huge overhead on a reader. We had such issues too.

So, I accepted his explanation about a motivation and assumed that a lot of small sstables == bad (which I still believe is the case) from then on. Hence my suggestion to compact them. 

If you believe that I'm missing something here and having thousands of small sstables is not going to cause any significant performance issue - then let's have this patch merged. 

But I think that @bhalevy was right and we need need to have protection against too many small sstables and their number can very quickly grow to 10s and even 100s of thousands.

And compaction that is going to have separate sstables reasonably large is going to represent such a protection in conjunction with a natural limit on how much data can be stored on a single node.

Does it make sense, @raphaelsc?

vladzcloudius commented:
> A large number of sstables doesn't cause an OOM, unless it's ridiculously large. A few thousand sstables should be fine.

Here (https://github.com/scylladb/scylladb/pull/14241#issuecomment-1590962805) @denesb tells the same thing Benny wrote before and I repeated after. We have to have a protection against too many small sstables and we already have a framework for that called "compaction". No need to invent anything new.

raphaelsc commented:
> > @vladzcloudius I don't understand why you think compaction of staging will make things any better. We're not deduplicating data, just reducing the sstable count. Could you please expand a bit more why you think that staging should be compacted? If that's to avoid the throttle slowing down repair, then it's a problem with throttling heuristic. We won't reshape the newly added sstables just to satisfy the existing weak heuristic.
> 
> According to @bhalevy the main reason of throttling was to avoid aggregation of too many small staging sstables. And we know that scylla doesn't handle such case well at all - there were multiple issues about that (and we were talking hundreds of thousands of sstables, @avikivity).
> 
> In addition having too many small sstables is indeed going to create a huge overhead on a reader. We had such issues too.
> 
> So, I accepted his explanation about a motivation and assumed that a lot of small sstables == bad (which I still believe is the case) from then on. Hence my suggestion to compact them.
> 
> If you believe that I'm missing something here and having thousands of small sstables is not going to cause any significant performance issue - then let's have this patch merged.
> 

Repair incrementally writes a sstable run into a table's maintainance set. We can read from this run as efficient as we can read from a single sstable. Due to disjointness, when scanning multiple keys, the reader can open a single fragment composing the run at a time (which is essential for memory efficiency). And for single-key reads, irrelevant fragments can be easily filtered out. The read and space amplification is the same between a run and a single large sstable. Otherwise, ICS wouldn't be feasible.

It seems to me the only reason you're suggesting to compact staging, is to reduce the sstable count, and therefore the today's queue size will allow view update backlog to be higher than what's allowed today, and then repair could go faster for a single table being repaired. As I understood it, it's adding unnecessary write amplification. If you think that view update throttler on repair should be relaxed, then it's the one that should be tweaked for allowing a higher backlog. But there are controversies around this approach too, but I'll leave that for @nyh who is the expert.

As for avoiding accumulation of small sstables, we can easily live with 256 (vnodes) * RF sstables until reshape happens, which is deferred until repair is done for a table.

bhalevy commented:
> Repair incrementally writes a sstable run into a table's maintainance set. We can read from this run as efficient as we can read from a single sstable. Due to disjointness, when scanning multiple keys, the reader can open a single fragment composing the run at a time (which is essential for memory efficiency). And for single-key reads, irrelevant fragments can be easily filtered out. The read and space amplification is the same between a run and a single large sstable. Otherwise, ICS wouldn't be feasible.

For that, let's start with https://github.com/scylladb/scylladb/issues/8478

denesb commented:
If we go ahead with this PR, we should fix https://github.com/scylladb/scylladb/issues/11744 (revive https://github.com/scylladb/scylladb/pull/11745). Currently view update generator takes all existing staging sstables in one go. If anything goes wrong, the entire progress is rolled back. With no backpressure anymore, repair freely piling on new sstables, this can mean that MV generation will possibly never finish.

bhalevy commented:
> As for avoiding accumulation of small sstables, we can easily live with 256 (vnodes) * RF sstables until reshape happens, which is deferred until repair is done for a table.

We have offstrategy compaction not only to bring those sstables into the compaction strategy invariant but also to reduce the number of overall sstables because with no back-pressure, say with a 100 shards, 100 tables, 100 shards, each creating 768 sstables you end with 7.68 million sstables, so if we remove the registration semaphore all together, repair should still need to be throttled and adjust its rate of based on the ability of downstream (view update generation and offstrategy compaction) to ingest the repaired sstables.  In a way, that's similar to #10954, except for the potential for keeping the repaired sstables as a run, for incremental selection.

asias commented:
> Let's reject repairs if there is backlog. Relying on the administrator is not deterministic.

We can either reject repair if there are too many sstables or block creation of repair_writer to generate new sstables. The first option is simpler and straightforward, but user / qa needs to restart the repair again and understand what is going on instead of report a bug that repair failed. The second option, user does not need to do anything.

asias commented:
> There's no point to compact disjoint sstables. It's just busy-work.

With more than one repair jobs, sstables from different run can be joint.  In this case, compaction is useful.

asias commented:
> @asias [fb95ff4](https://github.com/scylladb/scylladb/commit/fb95ff4efac8efa272a90a9a0ae1eb6c01a42963) this sounds like an important fix, but the commit message made it hard for me to understand the consequences exactly. could you please expand a bit more on that and tell me if it merits a backport?

We would pass wrong estimated number of partitions to sst->write_components().  How bad would it be?

asias commented:
> > A large number of sstables doesn't cause an OOM, unless it's ridiculously large. A few thousand sstables should be fine.
> 
> At the end of the day, disjoint sstables created by repair is no different than a sstable run. Do we compact a sstable run (which contains disjoint fixed-sized sstables) into a single larger sstable? no. we have mechanisms to prevent OOM, like incrementally reading from disjoint sstables. and we know that sstable metadata is proportional to the amount of partitions, and therefore the metadata is spread evenly across the fragments composing a run (unless we have a problem with partition estimation in repair).

Consider this:

repair job1: sst1 sst2 sst3 in staging 
repair job2: sst4 sst5 sst6 in staging
...


sst1,2,3 are disjoint, but sst1 and sst4 can be joint. 

Compaction helps in this case. 


> 
> It seems to me that the patch was created on the assumption that the the throttler is there to prevent an explosion on number of sstables, but I think the main goal is to allow view updates to keep up with new changes to base table brought in by repair.

denesb commented:
> > Let's reject repairs if there is backlog. Relying on the administrator is not deterministic.
> 
> We can either reject repair if there are too many sstables or block creation of repair_writer to generate new sstables. The first option is simpler and straightforward, but user / qa needs to restart the repair again and understand what is going on instead of report a bug that repair failed. The second option, user does not need to do anything.

Second option is the same as today. I also doubt option one is any more enticing for field-engineering.

asias commented:
> > > Let's reject repairs if there is backlog. Relying on the administrator is not deterministic.
> > 
> > 
> > We can either reject repair if there are too many sstables or block creation of repair_writer to generate new sstables. The first option is simpler and straightforward, but user / qa needs to restart the repair again and understand what is going on instead of report a bug that repair failed. The second option, user does not need to do anything.
> 
> Second option is the same as today. I also doubt option one is any more enticing for field-engineering.

Second is different from today: 1) new sstable is not created if throttled 2) the throttle is more accurate and direct with the knowledge of compaction manager

denesb commented:
> Second is different from today: 1) new sstable is not created if throttled 2) the throttle is more accurate and direct with the knowledge of compaction manager

The end result is the same: repair ends up stuck, not being able to create new sstables.

raphaelsc commented:
> > @asias [fb95ff4](https://github.com/scylladb/scylladb/commit/fb95ff4efac8efa272a90a9a0ae1eb6c01a42963) this sounds like an important fix, but the commit message made it hard for me to understand the consequences exactly. could you please expand a bit more on that and tell me if it merits a backport?
> 
> We would pass wrong estimated number of partitions to sst->write_components(). How bad would it be?

Were we overestimating or underestimating? Overestimating leads to bigger bloom filter, which can put a pressure on memory. Underestimating can lead to suboptimal bloom filter, which can impact read latency. Whatever it is, I think it merits a backport.

vladzcloudius commented:
> As for avoiding accumulation of small sstables, we can easily live with 256 (vnodes) * RF sstables until reshape happens, which is deferred until repair is done for a table.

@raphaelsc can we live with 10K sstables of 100 bytes each? 100K? 1M? 1M such such sstables is just ~100MB which is x10 less than a single typical ICS sstable. And repair can easily create you such a situation very quickly. 

And as @asias mentioned above all the "disjointness" argument is simply void.

raphaelsc commented:
> > As for avoiding accumulation of small sstables, we can easily live with 256 (vnodes) * RF sstables until reshape happens, which is deferred until repair is done for a table.
> 
> @raphaelsc can we live with 10K sstables of 100 bytes each? 100K? 1M? 1M such such sstables is just ~100MB which is x10 less than a single typical ICS sstable. And repair can easily create you such a situation very quickly.
> 
> And as @asias mentioned above all the "disjointness" argument is simply void.

I am getting confused as two different issues are being mixed. One is the view update throttler which slows down repair. Another is compaction not being triggered when overlapping is introduced to maintainance set. When that happens, it's an indication that reshape must be triggered regardless of whether the sstables live in staging or base directory (when there's no view involved). In this scenario, we compact to reduce read and space amplification, not satisfy the throttler.

vladzcloudius commented:
> > > As for avoiding accumulation of small sstables, we can easily live with 256 (vnodes) * RF sstables until reshape happens, which is deferred until repair is done for a table.
> > 
> > 
> > @raphaelsc can we live with 10K sstables of 100 bytes each? 100K? 1M? 1M such such sstables is just ~100MB which is x10 less than a single typical ICS sstable. And repair can easily create you such a situation very quickly.
> > And as @asias mentioned above all the "disjointness" argument is simply void.
> 
> I am getting confused as two different issues are being mixed. One is the view update throttler which slows down repair. Another is compaction not being triggered when overlapping is introduced to maintainance set. When that happens, it's an indication that reshape must be triggered regardless of whether the sstables live in staging or base directory (when there's no view involved). In this scenario, we compact to reduce read and space amplification, not satisfy the throttler.

We indeed mix both issues intentionally here because we believe one depends on another:

1) We want to change the throttler to not throttle the ongoing repair at all and instead wait for corresponding staging sstables to be fully processed **after** the actual repair of a base table is finished before declaring a repair as "done". This should allow repair finish as soon as possible eliminating a "traffic jam" phenomena. And at the same time provide the same consistency-after-repair guarantee as we have today with the throttler.
2) However we can't do (1) without resolving the "huge amount of small staging sstables" problem. And this can (and IMO should) be done by compacting staging sstables.

@raphaelsc

raphaelsc commented:
> > > > As for avoiding accumulation of small sstables, we can easily live with 256 (vnodes) * RF sstables until reshape happens, which is deferred until repair is done for a table.
> > > 
> > > 
> > > @raphaelsc can we live with 10K sstables of 100 bytes each? 100K? 1M? 1M such such sstables is just ~100MB which is x10 less than a single typical ICS sstable. And repair can easily create you such a situation very quickly.
> > > And as @asias mentioned above all the "disjointness" argument is simply void.
> > 
> > 
> > I am getting confused as two different issues are being mixed. One is the view update throttler which slows down repair. Another is compaction not being triggered when overlapping is introduced to maintainance set. When that happens, it's an indication that reshape must be triggered regardless of whether the sstables live in staging or base directory (when there's no view involved). In this scenario, we compact to reduce read and space amplification, not satisfy the throttler.
> 
> We indeed mix both issues intentionally here because we believe one depends on another:
> 
> 1. We want to change the throttler to not throttle the ongoing repair at all and instead wait for corresponding staging sstables to be fully processed **after** the actual repair of a base table is finished before declaring a repair as "done". This should allow repair finish as soon as possible eliminating a "traffic jam" phenomena. And at the same time provide the same consistency-after-repair guarantee as we have today with the throttler.
> 2. However we can't do (1) without resolving the "huge amount of small staging sstables" problem. And this can (and IMO should) be done by compacting staging sstables.

Today, we repair one table at a time, not one vnode range at a time anymore. To achieve 1, you don't need 2. If we want to defer view building (update) to after we're done repairing a table, that's very easy to implement and that won't cause the number of small sstables to explode. Then after we're done with view building, we will kick in off-strategy to integrate the staging sstables into the main set. The off-strategy kicks in after each repair completion on a table, is what prevents the accumulation of small sstables. Otherwise, we'd see an explosion of sstables when view update is not involved, as that throttler is not triggered when the base table being repaired doesn't have views.

vladzcloudius commented:
> Today, we repair one table at a time, not one vnode range at a time anymore. To achieve 1, you don't need 2. If we want to defer view building (update) to after we're done repairing a table, that's very easy to implement and that won't cause the number of small sstables to explode. 

Not sure I follow. Yes, we repair one table at a time and AFAIU when we repair a base table it's updates (in the context of repair) enter `staging` directory so that they trigger a corresponding MV update. And AFAIR we are going to flush these sstables as long as there is a first out-of-order key arriving. And this will result in very small sstables due to a nature of a repair process (since discrepancies usually not happening on joint tokens). And we see that these staging sstables are indeed very small.

Therefore if data is inconsistent enough (e.g. wasn't repaired for a long time) then the number of these sstables created during the whole repair can be very significant.

And according to aforementioned comment from Benny **this** was the main reason to have the throttler - so that the number of these sstables is kept in check. 

> The off-strategy kicking in after each repair completion on a table, is what prevents the accumulation of small sstables.

Hmmm... But what happens before the repair is over? You will have all those many tiny sstables and according to Botond above this **can** cause elevated memory consumption (and in the worst case - OOM). So, how do you resolve that (when there are not MVs)?

But if that somehow works then there is another intermediate quick fix: 
* not to replay staging sstables till the end of the repair.
* Run off-strategy repair on them without moving them to the main dir.
* Process the compaction result updating the MV.

Why this is probably not going to be a good final solution? - Because both the repair and staging processing times can be both very significant and we do want run both together.

raphaelsc commented:
> > Today, we repair one table at a time, not one vnode range at a time anymore. To achieve 1, you don't need 2. If we want to defer view building (update) to after we're done repairing a table, that's very easy to implement and that won't cause the number of small sstables to explode.
> 
> Not sure I follow. Yes, we repair one table at a time and AFAIU when we repair a base table it's updates (in the context of repair) enter `staging` directory so that they trigger a corresponding MV update. And AFAIR we are going to flush these sstables to a disk as soon as there is the first out-of-order key arriving. And this will result in very small sstables due to a nature of a repair process (since discrepancies usually not happening on joint tokens). And we see that these staging sstables are indeed very small.
> 
> Therefore if data is inconsistent enough (e.g. wasn't repaired for a long time) then the number of these sstables created during the whole repair can be very significant.
> 
> And according to aforementioned comment from Benny **this** was the main reason to have the throttler - so that the number of these sstables is kept in check.
> 
> > The off-strategy kicking in after each repair completion on a table, is what prevents the accumulation of small sstables.
> 
> Hmmm... But what happens before the repair is over? You will have all those many tiny sstables and according to Botond above this **can** cause elevated memory consumption (and in the worst case - OOM). So, how do you resolve that (when there are not MVs)?
> 
> But if that somehow works then there is another intermediate quick fix:
> 
> * not to replay staging sstables till the end of the repair.
> * Run off-strategy repair on them without moving them to the main dir.
> * Process the compaction result updating the MV.
> 
> Why this is probably not going to be a good final solution? - Because both the repair and staging processing times can be both very significant and we do want run both together.



> > Today, we repair one table at a time, not one vnode range at a time anymore. To achieve 1, you don't need 2. If we want to defer view building (update) to after we're done repairing a table, that's very easy to implement and that won't cause the number of small sstables to explode.
> 
> Not sure I follow. Yes, we repair one table at a time and AFAIU when we repair a base table it's updates (in the context of repair) enter `staging` directory so that they trigger a corresponding MV update. And AFAIR we are going to flush these sstables to a disk as soon as there is the first out-of-order key arriving. And this will result in very small sstables due to a nature of a repair process (since discrepancies usually not happening on joint tokens). And we see that these staging sstables are indeed very small.
> 
> Therefore if data is inconsistent enough (e.g. wasn't repaired for a long time) then the number of these sstables created during the whole repair can be very significant.

When reparing a table, we create RF sstables for each vnode range. Which is off-strategy has to work on ~768 sstables (RF=3) in a table when a new node is bootstrapped.

> 
> And according to aforementioned comment from Benny **this** was the main reason to have the throttler - so that the number of these sstables is kept in check.

That's a secondary effect. The reason for the throttler is to avoid overloading the view update queue, so the desynchronization between base and view table is bounded. If that's suboptimal, wrong, whatever, it's a different story. I found it by digging into old emails.

> 
> > The off-strategy kicking in after each repair completion on a table, is what prevents the accumulation of small sstables.
> 
> Hmmm... But what happens before the repair is over? You will have all those many tiny sstables and according to Botond above this **can** cause elevated memory consumption (and in the worst case - OOM). So, how do you resolve that (when there are no MVs)?

A node operation produces disjoint sstables, and off-strategy should compact that disjoint set as soon as the node operation is over. Asias provided an example of two repair jobs causing overlapping, and that's solved by kicking in off-strategy when overlapping is introduced to the maintaince sstable set (where new sstables written by repair goes).

> 
> But if that somehow works then there is another possible intermediate quick fix:
> 
> * not to replay staging sstables till the end of the repair.
> * Run off-strategy repair on them without moving them to the main dir.
> * Process the compaction result updating the MV.
> 
> Why this is probably not going to be a good final solution? - Because both the repair and staging processing times can be both very significant and we do want run both concurrently to save time.

why not serialize them? I initially thought you were suggesting to serialize repair and view building from staging. Running them concurrently is a guarantee we'll finish work faster? They're both cpu intensive and go through the same scheduling group. We're interested in throughput. 

Note we're kind of running them concurrently today. But repair completes first, as view building is much much slower. So repair gets blocked until view building completes its current round of work, to release semaphore units. Also note that even if we allow repair to always do work in parallel to view building, we'll still face terrible repair times as it must wait on view building completion. Repair time is still determined by how fast view building can get things done.

What's our goal here? Make repair faster when view building is involved. I don't see repair getting any faster unless we make the actual view building from staging faster.

vladzcloudius commented:
> When reparing a table, we create RF sstables for each vnode range. Which is off-strategy has to work on ~768 sstables (RF=3) in a table when a new node is bootstrapped.

We are not talking about bootstrapping (using RBNO) - we are talking about actual repair, @raphaelsc.

asias commented:
> > > @asias [fb95ff4](https://github.com/scylladb/scylladb/commit/fb95ff4efac8efa272a90a9a0ae1eb6c01a42963) this sounds like an important fix, but the commit message made it hard for me to understand the consequences exactly. could you please expand a bit more on that and tell me if it merits a backport?
> > 
> > 
> > We would pass wrong estimated number of partitions to sst->write_components(). How bad would it be?
> 
> Were we overestimating or underestimating? Overestimating leads to bigger bloom filter, which can put a pressure on memory. Underestimating can lead to suboptimal bloom filter, which can impact read latency. Whatever it is, I think it merits a backport.

 It is underestimating.  Yes, we need a backport.

denesb commented:
> > That's a secondary effect. The reason for the throttler is to avoid overloading the view update queue, so the desynchronization between base and view table is bounded. If that's suboptimal, wrong, whatever, it's a different story. I found it by digging into old emails.
> 
> I'm not sure what's more important after we are eventual consistent and views are eventually in-sync too.
> 
> What is clear however is that this throttler creates a "traffic jam" effect causing repair last for more than a week where it is supposed to last not more than 24h.

I think that if we have to choose between an inconsistent base + inconsistent views and a consistent base (repaired) + inconsistent views, the latter is better. We are not making anything worse by detaching the base repair from the progress of the quasi view repair. All this is to say, I do see the value in this PR and I think it is a good idea. I just want to make sure we have some mechanism to prevent *unbounded* accumulation of staging sstables. I don't think there is a problem with accumulating staging sstables in general. 99.9% of clusters probably would have smooth sailing even after this PR. But I don't want to be woken up in the middle of the night, because that 0.01% hit a runaway staging sstable count and now the cluster is crashing and nodes won't start up, instant OOMing due to number of sstables (we've seen cases like this).

asias commented:
Changes in v2:
- Updated b/test/boost/view_build_test.cc
- Updated the commit message
- Be more conservative on the max pending sstables in staging dir per shard, 100. Since we do not compact sstables in staging dir for now.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2062/

denesb commented:
PR reverted in 4cee8206f860ab776e570d45ee512045995e3fa9.
The test modified by the PR view_build_test.test_view_update_generator_register_semaphore_unit_leak breaks the build often.
See e.g. https://jenkins.scylladb.com/job/scylla-master/job/build/1533/

denesb commented:
Yes, I guess we could re open. But AFAIK we won't be able to merged a re-opened PR. So a new one will have to be opened. But I might remember wrong.

asias commented:
> PR reverted in [4cee820](https://github.com/scylladb/scylladb/commit/4cee8206f860ab776e570d45ee512045995e3fa9). The test modified by the PR view_build_test.test_view_update_generator_register_semaphore_unit_leak breaks the build often. See e.g. https://jenkins.scylladb.com/job/scylla-master/job/build/1533/

Botond,  I saw you wrote the test test_view_update_generator_register_semaphore_unit_leak in commit 22a64937163252f23fb68d3147d65f0f2c5cdc35. 

I am not sure why we have 

`20:46:03  unknown location(0): fatal error: in "test_view_update_generator_register_semaphore_unit_leak": seastar::broken_named_semaphore: Semaphore broken: view update generator`

Could you help to take a look?


```
20:46:03  INFO  2023-07-13 12:29:23,286 [shard 1] database - Joining memtable update action
20:46:03  INFO  2023-07-13 12:29:23,479 [shard 0] compaction_manager - Asked to stop
20:46:03  INFO  2023-07-13 12:29:23,480 [shard 0] task_manager - Stoppping module compaction
20:46:03  INFO  2023-07-13 12:29:23,481 [shard 0] task_manager - Unregistered module compaction
20:46:03  INFO  2023-07-13 12:29:23,483 [shard 1] compaction_manager - Asked to stop
20:46:03  INFO  2023-07-13 12:29:23,483 [shard 0] compaction_manager - Stopped
20:46:03  INFO  2023-07-13 12:29:23,483 [shard 1] task_manager - Stoppping module compaction
20:46:03  INFO  2023-07-13 12:29:23,485 [shard 1] task_manager - Unregistered module compaction
20:46:03  INFO  2023-07-13 12:29:23,486 [shard 1] compaction_manager - Stopped
20:46:03  unknown location(0): fatal error: in "test_view_update_generator_register_semaphore_unit_leak": seastar::broken_named_semaphore: Semaphore broken: view update generator
20:46:03  test/boost/view_build_test.cc(657): last checkpoint
20:46:03  test/boost/view_build_test.cc(585): Leaving test case "test_view_update_generator_register_semaphore_unit_leak"; testing time: 320060787us
20:46:03  test/boost/view_build_test.cc(699): Test case "test_view_update_generator_buffering" is skipped because disabled
20:46:03  test/boost/view_build_test.cc(878): Test case "test_load_view_build_progress_with_values_missing" is skipped because disabled
20:46:03  test/boost/view_build_test.cc(893): Test case "test_view_update_generator_buffering_with_random_mutations" is skipped because disabled
20:46:03  Leaving test module "Master Test Suite"; testing time: 320060958us
```

denesb commented:
> Botond, I saw you wrote the test test_view_update_generator_register_semaphore_unit_leak in commit [22a6493](https://github.com/scylladb/scylladb/commit/22a64937163252f23fb68d3147d65f0f2c5cdc35).
> 
> I am not sure why we have
> 
> `20:46:03 unknown location(0): fatal error: in "test_view_update_generator_register_semaphore_unit_leak": seastar::broken_named_semaphore: Semaphore broken: view update generator`
> 
> Could you help to take a look?

I think the problem is simply that with the new increased limits, the test creates more sstables and it now doesn't finish before the semaphore is shut down. The test might need restructuring to work with the new elevated limits.

asias commented:
> > Botond, I saw you wrote the test test_view_update_generator_register_semaphore_unit_leak in commit [22a6493](https://github.com/scylladb/scylladb/commit/22a64937163252f23fb68d3147d65f0f2c5cdc35).
> > I am not sure why we have
> > `20:46:03 unknown location(0): fatal error: in "test_view_update_generator_register_semaphore_unit_leak": seastar::broken_named_semaphore: Semaphore broken: view update generator`
> > Could you help to take a look?
> 
> I think the problem is simply that with the new increased limits, the test creates more sstables and it now doesn't finish before the semaphore is shut down. The test might need restructuring to work with the new elevated limits.

Btw, I could not produce locally.

denesb commented:
https://github.com/scylladb/scylladb/pull/14727 was merged, so this PR can be reinstated.
@asias please make sure the test passes, before re-submitting the PR.

asias commented:
> #14727 was merged, so this PR can be reinstated. @asias please make sure the test passes, before re-submitting the PR.

Rebased. It passed with debug mode.

asias commented:
@denesb  New PR is here: https://github.com/scylladb/scylladb/pull/14750
